psutil/_psutil_bsd.c

```diff
@@ -391,6 +396,104 @@ psutil_proc_cmdline(PyObject *self, PyObject *args) {
 }
 
 
+/*
+ * Return process environment as a Python dictionary
+ */
+PyObject *
+psutil_proc_environ(PyObject *self, PyObject *args) {
+    unsigned int i, cnt = 0;
+    long pid;
+    char **envs;
+    char errbuf[_POSIX2_LINE_MAX];
+    kvm_t *kd;
+#ifdef PSUTIL_NETBSD
+    struct kinfo_proc2 *p;
+#else
+    struct kinfo_proc *p;
+#endif
+    PyObject *py_retdict;
+
+    if (!PyArg_ParseTuple(args, "l", &pid))
+        return NULL;
+
+#if defined(PSUTIL_FREEBSD)
+    kd = kvm_openfiles(NULL, "/dev/null", NULL, 0, errbuf);
+#else
+    kd = kvm_openfiles(NULL, NULL, NULL, KVM_NO_FILES, errbuf);
+#endif
+    if (!kd) {
+        PyErr_SetString(PyExc_RuntimeError, errbuf);
+        return NULL;
+    }
+
+#if defined(PSUTIL_FREEBSD)
+    p = kvm_getprocs(kd, KERN_PROC_PID, pid, (int *)&cnt);
+#elif defined(PSUTIL_OPENBSD)
+    p = kvm_getprocs(kd, KERN_PROC_PID, pid, sizeof(*p), &cnt);
+#elif defined(PSUTIL_NETBSD)
+    p = kvm_getproc2(kd, KERN_PROC_PID, pid, sizeof(*p), &cnt);
+#endif
+
+    if (!p)
+    {
+        PyObject *ret = NoSuchProcess(kvm_geterr(kd));
+        kvm_close(kd);
+        return ret;
+    }
+
+    if (!cnt)
+    {
+        kvm_close(kd);
```

Comment:
I suggest to put all these `kvm_close(kd);` in the `error` section below, and do `goto error;` here.